File: lib/opencoinage/issuer.rb
module OpenCoinage
  ##
  # A digital currency issuer.
  class Issuer
    RDF_TYPE = Vocabulary[:Issuer]

    ##
    # Initializes the issuer.
    #
    # @param  [RDF::URI, String, #to_s] uri
    #   a valid URI identifying the issuer
    # @param  [Hash{Symbol => Object}]  options
    # @option options [String] :name (nil)
    #   the name of the issuer
    def initialize(uri = nil, options = {})
      @uri  = RDF::URI(uri || UUID.generate(:urn))
      @name = options[:name] if options.has_key?(:name)
    end

    ##
    # The URI identifying this issuer.
    #
    # @return [RDF::URI]
    attr_reader :uri
    alias_method :to_uri, :uri

    ##
    # The name of this issuer.
    #
    # @return [String]
    # @see    http://xmlns.com/foaf/spec/#term_name
    attr_accessor :name

    ##
    # Returns the hash table representation of this issuer.
    #
    # @return [Hash]
    def to_hash
      {:name => name}
    end

    ##
    # Returns the JSON representation of this issuer.
    #
    # @return [String]
    # @see    http://en.wikipedia.org/wiki/JSON
    # @see    http://json.rubyforge.org/
    def to_json
      require 'json' unless defined?(::JSON)
      to_hash.to_json
    end

    ##
    # Returns the RDF representation of this issuer.
    #
    # @param  [Hash{Symbol => Object}] options
    # @option options [RDF::Resource]  :context (nil)
    # @return [RDF::Graph]
    # @see    http://en.wikipedia.org/wiki/Resource_Description_Framework
    # @see    http://rdf.rubyforge.org/RDF/Graph.html
    def to_rdf(options = {})
      RDF::Graph.new(options[:context]) do |graph|
        graph << [uri, RDF.type, self.class.const_get(:RDF_TYPE)]
        graph << [uri, RDF::FOAF.name, name] if name
      end
    end
  end # Issuer
end # OpenCoinage

File: src/java/org/opencoinage/Issuer.java
package org.opencoinage;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Map;

/**
 * A digital currency issuer.
 */
public class Issuer {
  public static final String RDF_TYPE = Vocabulary.BASE + "Issuer";

  /**
   * The URI identifying this issuer.
   *
   * @see java.net.URI
   */
  public final URI uri;

  /**
   * The name of this issuer.
   *
   * @see http://xmlns.com/foaf/spec/#term_name
   */
  public String name;

  /**
   * @param uri a valid URI identifying the issuer
   */
  public Issuer(final URI uri) {
    this.uri = uri;
  }

  /**
   * @param uri a valid URI identifying the issuer
   * @param options key-value mappings for fields
   */
  public Issuer(final URI uri, final Map<String, ?> options) {
    this.uri = uri;
    if (options != null) {
      if (options.containsKey("name")) {
        this.name = (String)options.get("name");
      }
    }
  }

  /**
   * @param uri a valid URI identifying the issuer
   * @throws URISyntaxException if <code>uri</code> is not a valid URI
   */
  public Issuer(final String uri) throws URISyntaxException {
    this(new URI(uri));
  }

  /**
   * @param uri a valid URI identifying the issuer
   * @param options key-value mappings for fields
   * @throws URISyntaxException if <code>uri</code> is not a valid URI
   */
  public Issuer(final String uri, final Map<String, ?> options) throws URISyntaxException {
    this(new URI(uri), options);
  }
}
